src/modules/lua.c

```diff
@@ -58,6 +58,79 @@ static pthread_mutex_t mtev_lua_states_lock = PTHREAD_MUTEX_INITIALIZER;
 static mtev_hash_table mtev_coros;
 static pthread_mutex_t coro_lock = PTHREAD_MUTEX_INITIALIZER;
 
+struct lua_module_gc_params {
+  int iters_since_full;
+  int full_every;
+  int steps;
+  int steps_multiplier, set_steps_multiplier;
+  int pause_size, set_pause_size;
+};
+
+static lua_module_gc_params_t default_gc_params = {
+  .iters_since_full = 0,
+  .full_every = 1000,
+  .steps = 0,
+  .steps_multiplier = 1, .set_steps_multiplier = 1,
+  .pause_size = 200, .set_pause_size = 200
+};
+
+lua_module_gc_params_t *
+mtev_lua_config_gc_params(mtev_hash_table *o) {
+  lua_module_gc_params_t *p = calloc(1, sizeof(*p));
+  memcpy(p, &default_gc_params, sizeof(*p));
+  if(o) {
+    const char *str;
+    if(mtev_hash_retr_str(o, "gc_full", strlen("gc_full"), &str))
+      p->full_every = atoi(str);
+    if(mtev_hash_retr_str(o, "gc_step", strlen("gc_step"), &str))
+      p->steps = atoi(str);
+    if(mtev_hash_retr_str(o, "gc_pause", strlen("gc_pause"), &str))
+      p->set_pause_size = atoi(str);
+    if(mtev_hash_retr_str(o, "gc_stepmul", strlen("gc_stepmul"), &str))
+      p->set_steps_multiplier = atoi(str);
+  }
+  return p;
+}
+
+void
+mtev_lua_set_gc_params(lua_module_closure_t *lmc, lua_module_gc_params_t *p) {
+  if(lmc->gcparams == NULL) {
+    lua_module_gc_params_t *newp = calloc(1, sizeof(*newp));
+    memcpy(newp, p, sizeof(*p));
+    lmc->gcparams = newp;
+  }
+  else {
+    memcpy(lmc->gcparams, p, sizeof(*p));
+  }
+}
+
+void
+mtev_lua_gc(lua_module_closure_t *lmc) {
+  lua_State *L = lmc->lua_state;
+  lua_module_gc_params_t *p = lmc->gcparams;
+  if(!p) {
+    lua_gc(L, LUA_GCCOLLECT, 0);
+    return;
+  }
+
+  if(p->steps_multiplier != p->set_steps_multiplier) {
+    lua_gc(L, LUA_GCSETSTEPMUL, p->set_steps_multiplier);
+    p->steps_multiplier = p->set_steps_multiplier;
+  }
+  if(p->pause_size != p->set_pause_size) {
+    lua_gc(L, LUA_GCSETPAUSE, p->set_pause_size);
+    p->pause_size = p->set_pause_size;
+  }
+
+  p->iters_since_full++;
+  if(p->full_every && (p->iters_since_full > p->full_every)) {
+    p->iters_since_full = 0;
+    lua_gc(L, LUA_GCCOLLECT, 0);
+    return;
+  }
+  lua_gc(L, LUA_GCSTEP, p->steps);
```

Comment:
Check if `p->steps > 0` before calling lua_gc here?

Reply:
No, we want to call with steps == 0, that's the default.